README
README,v 2.3 1994/02/21 00:04:07 dmoore Exp

*** General:

This is a BETA release.  The server should be stable, but there currently
isn't much admin documentation.  Player documentation is available
online via 'help' and 'man'.  A player tutorial is available in
docs/Muck.man.txt.

You may want to look at the scripts in admin/scripts after you've
played with your server for a while.  There is a sample restart
script in the minimal-db directory which you may want to use.

IMPORTANT: if you are upgrading from 2.3pl2 to this version, your
	   database needs to be upgraded.  Use the provided script:
		./convert-pl2-to-pl3 ~/mud/tt/data/in.tt.new
	   For example.  It will make a backup of your database
	   in <db>.bak where <db> is the name you gave it.
	   Run the new sanity checker on this database to see how
	   it is.

IMPORTANT: if you are upgrading from 2.2 or the supported 2.2 derivatives
	   you must use the convert _program_ on your database.


*** Mailing list:

There is a mailing list for tinymuck servers at:

tinymuck-sloggers@tcp.com

To join the list send mail to:

majordomo@tcp.com with `subscribe tinymuck-sloggers' in the body.

It is recommended that all people trying the server join the mailing list.
This way you can be informed of new releases, get help and suggestions
for administration issues (like: 'aiee, my db seems to be corrupted' or
'hmm, what is a good way of purging old players).



*** Reporting bugs:

The author can be reached via dmoore@ucsd.edu.  Please include the string
'tmuck2.4' in the subject to ensure faster response to letters.  Send all
bug reports to this address.  An example of a good subject for a bug report
would be "fatal @open bug in tmuck2.4".  A bad subject would be "yo, ojay".

Include:

1.	E-Mail Address.
2.	What causes it?
3.	Where in source code, if known?
4.	Suggested fix, if you have one.



*** Deriviative works:

1. Should you make a substantial addition to this work and choose to
distribute it, please inform me first.  If that is not possible, in
your distribution, you should either:

	a) Not use the name TinyMUCK as the program, and note and
	acknowledge it's deriviative nature.

	b) Use the name TinyMUCK and it's version number, but append
	a suffix to distinguish it from the original.
	(Eg. tinymuck2.2fb is the name used by the fuzzball deriviative.)

2. You should distribute the source with the binaries (if binaries are
provided), and update the documentation to reflect the changes you
made.  If you are unable to do the latter, you must state the changes
in some way evident to the user when either running the software or
when unpacking the archive.

3. Inform me of any note-worthy uses of this software.



*** Thanks, and enjoy

David Moore


CHANGES
CHANGES,v 2.50 1997/08/30 07:21:29 dmoore Exp

Changes are now semi-sorted into the categories of player, wizard, admin,
and programmer.  The player and wizard lists are things that might affect
those people.  The admin and programmer sections have some overlap, and
my criterion for the division is that admin items are usually normal
header file settings.  Programmer section contains information about
changes to more internal structures.

2.4:
----
PLAYER:
o	New strftime formats %i and %J.
o	Players are now notified when they are forced by a wizard.
o	In the case of a remote kill, both the room of the killer and the
	killee are notified of the kill.
WIZARD:
ADMIN:
o	New script 'admin/scripts/findflag' which prints out dbrefs
	having a given flag.  Ex: findflag DARK in.foo.new
PROGRAMMER:
o	Changed the muck_strftime to take an extra argument to
	let you use GMT time instead of the default of local time.
o	Added %i to muck_strftime which returns a value like the
	idle time for the WHO list.
o	Added %J to muck_strftime which is the same as %j, except it
	uses leading spaces instead of leading 0s.  This is used for
	connect time in the WHO list.
o	Changed the connect time/idle time for WHO list to use
	Bufsprint with %T.
o	Fixed typo in log.c in log_last_command.
o	Added HAVE_STRERROR to disable setting of strerror macro,
	instead of old checking for an already existing macro by that
	name.
o	Added HAVE_TZSET for people who have it.  Now you could
	change the timezone of the mud by just setting an environment
	variable TZ (perhaps in your restart script).
o	Added HAVE_TZNAME for places (sys5) which don't have the
	time zone in the struct tm, but do have it in tzname.
o	Fixed up a call to strcasecmp and 3 to strncasecmp which showed
	up incorrectly.  Changed to muck_stricmp, and muck_strnicmp.
	Removed HAVE_STRCASECMP.
o	Added ALIGN_BYTES to tell muck_malloc and debug_malloc about
	alignment requirements.
o	The tracing option for debugging malloc now checks the magic
	numbers for the unfreed resources, to detect any overruns
	they may have had.
o	Fixed more getdtable size portability.
o	Check more signals for existance on particular system before
	using them.
o	Fixed bug in compiler which would allow a tab to be placed in
	a string.
o	Fixed a bug with code structures where the refcount was decremented
	inappropriately if a program had a ' construct and failed to
	compile.  Removed clear_inst and clear_inst_un.  Now use
	clear_inst_code if you are nuking an inst which is in a code
	struct (ie clear_code or in compiler).  Use clear_inst_interp
	if you are nuking an inst that is on a muf stack or similar
	(ie anything in the interp).
o	Fixed a bug with failed forks preventing future forked dumps.
o	A failed 'accept' call will no longer cause the server to exit,
	seen in the logs as: 'PERROR: main_loop: new_connection'.
	Now it will be logged, but the server continues running as before.
o	Added log_sanity command, for use by sanity routines.
o	On systems without 'random', use 3 calls to rand to hopefully get
	31 bits of a random number.
o	When doing batch input processing, kick out of the loop to allow
	output at reasonable intervals to keep perceived interactivity
	level higher.
o	Added a send_except call to interface.c to allow for more
	efficient implementation of notify_except.
o	Fixed bug which gave people 1 penny for every unlinked exit they
	owned, anytime anyone did a recycle.
o	Added make2_shared_string to speed up prim_strcat.
o	Sped up property hashtable comparision, to avoid a stricmp call.
o	Changed con_info in interface.c to return the time of the last
	player command, not the idle time, since calling time() can be
	very slow.  Now prim_conidle calls time() itself when needed.
o	Added a match_everything call to match.c and made some routines
	use it (more should be done).  Improved ordering of operations in
	match_everything provides significant speedups for .pmatch and
	similar operations in rooms with many objects.
o	The muf `exit' primitive has been replaced with INST_EXIT.
o	A new instruction INST_PICK acts like pick+1 and is used for dup,
	swap, and pick by the optimizer when the operation is guaranteed
	to succeed.
o	New instruction INST_WIZCHECK forces a check for the state of
	the frame to have wizard permissions.  It is emitted by the
	optimizer in front of primitives which require wizperms.
o	New file compile/optimize.c implements compile time stack depth
	and argument type checking.  Other small optimizations are done
	including some dead code elimination and minor constant folding.
o	Various changes to the interpreter to speed it up quite a bit.
o	The split bytecode and bytedata has been unified again.  The
	space savings were insignificant.
o	The muf flag? primitive now respects QUELL when checking for
	the GOD flag (when GOD_FLAGS are enabled).
	

2.3 beta patchlevel 3:
----------------------
PLAYER:
o	@stat, @find, @owned are now free for players.

o	Server page costs PAGE_COST not LOOKUP_COST now.  PAGE_COST
	defaults to 0.  If a player has negative pennies, they can't page.

o	Object matching via '*player' is always free.

o	Made @stat return global mud information for all players, and allow
	any player to @stat themselves.

o	Whispering to an object now yields an error rather than just
	saying that the object is not connected.

o	Added a magic cookie for '#' -> 'poundsign'.

o	Programs no longer have that default 'scroll' message for a desc.

o	Run programs on #0 and player on connect and disconnect.  Look
	in properties named _connect and _disconnect.  Acts like @desc,
	in that you'd do 'message' or '@545 arguments', etc.  If you use
	MUF_CONNECT_HOOKS.

WIZARD:

ADMIN:
o	No more config.h in distribution.  Copy include/config.h-dist
	to include/config.h and edit that.  The files config.h-temp
	and config.h-druid contain samples of a different way
	of handling config.h.

o	Portability change: added HAVE_UNISTD_H in config.h.

src/db/flags.c
/* Copyright (c) 1992 by David Moore.  All rights reserved. */
/* flags.c,v 2.8 1994/02/21 08:32:48 dmoore Exp */
#include "config.h"

#include "db.h"
#include "params.h"
#include "flags.h"
#include "buffer.h"
#include "externs.h"

struct object_type_name {
    const char *long_name;
    const char *short_name;
    object_flag_type kind;
};

static struct object_type_name type_names [] = {
    { "ROOM",           "R" },
    { "THING",          ""  },
    { "EXIT/ACTION",    "E" },
    { "PLAYER",         "P" },
    { "FORTH",          "F" },
    { "---",            ""  },
    { "GARBAGE",        ""  },
    { "NOTYPE",         ""  },
    { NULL, NULL }
};

struct object_flag_name {
    const char *long_name;
    const char *short_name;
    object_flag_type flag;
    object_flag_type kind;
};

static struct object_flag_name flag_names[] = {
#ifdef DRUID_MUCK
    { "INVISIBLE","I", INVISIBLE, TYPE_PLAYER },
#endif
#ifdef GOD_FLAGS
    { "GOD",	  "G", GOD,	 TYPE_ANY },
#endif
    { "WIZARD",   "W", WIZARD,   TYPE_ANY },
    { "LINK_OK",  "L", LINK_OK,  TYPE_ANY },
    { "DARK",     "D", DARK,     TYPE_ROOM },
    { "DARK",     "D", DARK,     TYPE_PLAYER },
    { "DARK",     "D", DARK,     TYPE_THING },
    { "DARK",     "D", DARK,     TYPE_EXIT },
    { "DEBUG",    "D", DEBUG,    TYPE_PROGRAM },
    { "STICKY",   "S", STICKY,   TYPE_ROOM },
    { "STICKY",   "S", STICKY,   TYPE_PLAYER },
    { "STICKY",   "S", STICKY,   TYPE_THING },
    { "STICKY",   "S", STICKY,   TYPE_EXIT },
    { "SETUID",   "S", SETUID,   TYPE_PROGRAM },
    { "BUILDER",  "B", BUILDER,  TYPE_ANY },
    { "CHOWN_OK", "C", CHOWN_OK, TYPE_ANY },
    { "JUMP_OK",  "J", JUMP_OK,  TYPE_ANY },
    { "HAVEN",    "H", HAVEN,    TYPE_ANY },
    { "ABODE",    "A", ABODE,    TYPE_ANY },
    { "MURKY",    "M", MURKY,    TYPE_ROOM},
    { "MUCKER",   "M", MUCKER,   TYPE_PLAYER },
    { "MUCKER",   "M", MUCKER,   TYPE_THING },
    { "MUCKER",   "M", MUCKER,   TYPE_EXIT },
    { "MUCKER",   "M", MUCKER,   TYPE_PROGRAM },
    { "QUELL",    "Q", QUELL,    TYPE_ANY },
    { "LINE_NUMBERS",  "",  LINE_NUMBERS, TYPE_PLAYER },
    { NULL, NULL, 0, 0 }
};


const char *unparse_flag(const dbref thing, object_flag_type flag, const int long_name)
{
    static Buffer buf;
    struct object_flag_name *temp;

    Bufcpy(&buf, "");
    for (temp = flag_names; temp->long_name; temp++) {
	if ((flag & temp->flag)
            && ((temp->kind == TYPE_ANY) || (Typeof(thing) == temp->kind))) {
	    if (long_name) Bufcat(&buf, temp->long_name);
	    else Bufcat(&buf, temp->short_name);
	}
    }

    return Buftext(&buf);
}


const char *unparse_flags(const dbref thing)
{
    static Buffer buf;
    struct object_flag_name *temp;

    /* Stick in the type, P, F, E, etc. */
    Bufcpy(&buf, type_names[Typeof(thing)].short_name);

    /* Loop building up the other flags. */
    for (temp = flag_names; temp->long_name; temp++) {
	if (HasFlag(thing, temp->flag)
	    && ((temp->kind == TYPE_ANY)
		|| (Typeof(thing) == temp->kind))) {
	    /* This object has both this flag set, and is the right type. */
	    Bufcat(&buf, temp->short_name);
	}
    }

    return Buftext(&buf);
}


const char *unparse_long_flags(const dbref thing, const int dump_format)
{
    struct object_flag_name *temp;
    int saw_a_flag = 0;
    static Buffer buf;

    /* Stick in the type. */
    if (!dump_format) {
	Bufcpy(&buf, "Type: ");
    } else {
	Bufcpy(&buf, "");
    }
    Bufcat(&buf, type_names[Typeof(thing)].long_name);

    /* Loop building up the other flags. */
    for (temp = flag_names; temp->long_name; temp++) {
	if (HasFlag(thing, temp->flag)
	    && ((temp->kind == TYPE_ANY)
		|| (Typeof(thing) == temp->kind))) {
	    /* This object has both this flag set, and is the right type. */
	    if (!dump_format && !saw_a_flag) {
		Bufcat(&buf, "  Flags:");
		saw_a_flag++;
	    }
	    Bufcatlist(&buf, " ", temp->long_name, NULL);
	}
    }

    return Buftext(&buf);
}


object_flag_type parse_flags(const char *flagstr)
{
    object_flag_type objtype, result;
    const char *p;
    struct object_flag_name *tflag;
    struct object_type_name *ttype;
    Buffer buf;

    result = 0;
    
    if (!flagstr) return 0;

    /* Copy out the type of the object. */
    Bufcpy(&buf, "");
    for (p = flagstr; *p && (*p != ' '); p++)
	Bufcat_char(&buf, *p);
   
    /* Move up flagstr to be just after found type. */
    if (*p) flagstr = p + 1;
    else flagstr = p;

    /* Lookup the type. */
    p = Buftext(&buf);
    for (ttype = type_names; ttype->long_name; ttype++) {
	if (!muck_stricmp(p, ttype->long_name)) {
	    objtype = result = ttype - type_names;
	    break;
	}
    }

    /* Loop collecting all of the other flags. */
    while (*flagstr) {
	Bufcpy(&buf, "");
	for (p = flagstr; *p && (*p != ' '); p++)
	    Bufcat_char(&buf, *p);

	/* Move up flagstr to just after found flag. */
	if (*p) flagstr = p + 1;
	else flagstr = p;

	/* Lookup the flag. */
	p = Buftext(&buf);
	for (tflag = flag_names; tflag->long_name; tflag++) {
	    if ((!muck_stricmp(p, tflag->long_name))
		&& ((tflag->kind == TYPE_ANY)
		    || (objtype == tflag->kind))) {
		result |= tflag->flag;
		break;
	    }
	}
    }

    return result;
}


/* Lookup the flag for a given object and string version of flag name.
   If no such flag return BAD_FLAG (-1). */
int which_flag(object_flag_type type, const char *flagstr)
{
    struct object_flag_name *temp;    

    /* Loop building up the other flags. */
    for (temp = flag_names; temp->long_name; temp++) {
	if ((temp->kind != TYPE_ANY) && (temp->kind != type))
	    continue;

	if (muck_strprefix(flagstr, temp->short_name)) {
	    if (muck_strprefix(temp->long_name, flagstr)) {
		return temp->flag;
	    } else {
		/* Matched beginning of a valid flag, but remainder is bad. */
		return BAD_FLAG;
	    }
	}
    }

    return BAD_FLAG;		/* Didn't end up matching anything. */
}
